File: app/src/main/java/com/shekar/alamomvp/ui/base/BaseViewModel.java
package com.shekar.alamomvp.ui.base;

import android.databinding.BaseObservable;
import rx.subscriptions.CompositeSubscription;
import timber.log.Timber;

public abstract class BaseViewModel<T extends MvvmView> extends BaseObservable
    implements MvvmViewModel<T> {

  protected CompositeSubscription compositeSubscription;
  private T mvvmView;

  @Override
  public void attachView(T mvvmView) {
    this.mvvmView = mvvmView;
  }

  @Override
  public void detachView() {
    mvvmView = null;
  }

  public boolean isViewAttached() {
    return mvvmView != null;
  }

  public T getMvvmView() {
    return mvvmView;
  }

  public void checkViewAttached() {
    if (!isViewAttached()) throw new MvpViewNotAttachedException();
  }

  public void unsubscribeFromDataStore() {
    Timber.d("unsubscribeFromDataStore(): ");
    if(compositeSubscription!=null) {
      compositeSubscription.unsubscribe();
      compositeSubscription.clear();
      compositeSubscription = null;
    }
  }

  public static class MvpViewNotAttachedException extends RuntimeException {
    public MvpViewNotAttachedException() {
      super("Please call Presenter.attachView(CategoryMvpView) before" +
          " requesting data to the Presenter");
    }
  }
}

File: app/src/main/java/com/shekar/alamomvp/ui/base/BaseViewModel.kt
package com.shekar.alamomvp.ui.base

import android.databinding.BaseObservable
import rx.subscriptions.CompositeSubscription
import timber.log.Timber

abstract class BaseViewModel<T : MvvmView> : BaseObservable(), MvvmViewModel<T> {

  protected var compositeSubscription: CompositeSubscription? = null
  var mvvmView: T? = null

  override fun attachView(mvvmView: T) {
    this.mvvmView = mvvmView
  }

  override fun detachView() {
    mvvmView = null
  }

  val isViewAttached: Boolean
    get() = mvvmView != null

  fun checkViewAttached() {
    if (!isViewAttached) throw MvpViewNotAttachedException()
  }

  open fun unsubscribeFromDataStore() {
    Timber.d("unsubscribeFromDataStore(): ")
    if (compositeSubscription != null) {
      compositeSubscription!!.unsubscribe()
      compositeSubscription!!.clear()
      compositeSubscription = null
    }
  }

  class MvpViewNotAttachedException : RuntimeException(
      "Please call Presenter.attachView(CategoryMvpView) before" + " requesting data to the Presenter")
}